### PR TITLE
Transient per-turn state with decay, and promotion of reinforced state

### LINKED ISSUE #120: Transient per-turn field state (mood/intent/posture) with provenance + decay
Part of the **Extraction, Continuity & State** milestone. From the pre-rebuild target feature list:

> Transient per-turn field state (mood/intent/posture) with provenance + decay

## Current state
**⚠️ Pre-rebuild concept — needs re-scoping.** The original design assumed a per-turn extraction pipeline; the rebuilt app extracts once per scene (`POST .../absorb`) and a chat turn makes exactly one LLM call (`routes.post_chat` → `context.build_messages` → `_chat_stream`).

What exists is a per-NPC *standing* snapshot: `backend/src/grimoire/store/playstate.py` (`characters/<cid>/state.md`, `## Current state / ## Knows / ## Suspects` prose), written only via absorb StagedEdits (`store/absorb.py:materialize`/`apply_edits`) and injected via `context._character_states` + `templates/scene/sections/character_state.j2`. It has scene granularity, no provenance, no decay. Posts carry no metadata — a transcript is `**Speaker:** content` blocks (`store/scenes.py`); a post's identity is its index.

Be honest about cost: live per-turn mood/intent/posture needs either (a) a second LLM pass after every turn (doubling calls/latency/spend through the sole `OpenRouterClient`), or (b) an inline-tracker output mode where the reply carries a machine-readable block — which is exactly #109's inline mode.

### Option A — inline tracker block (recommended, alongside #109)
- Prompt (template under `templates/scene/`, never hard-coded) instructs the model to end each reply with a fenced JSON block `{"state": {"<actor>": {"mood": "...", "intent": "...", "posture": "..."}}}`.
- `routes._persist_reply` strips/parses the block before `scenes.split_reply`, so it never enters the transcript.
- Store `campaigns/<cid>/turnstate.json`: `{sid: {post_index: {actor_token: {field: value}}}}` — provenance is the (sid, post index).
- Decay: at context build, inject only entries whose post index is within N posts of the tail (N via `store/config.py`, default e.g. 4); older values simply drop out.
- New section template + `_SECTIONS` entry in `store/context.py` so it appears in the `GET .../context` token breakdown.
- Recommended: zero extra LLM calls; the model was going to narrate these cues anyway.

### Option B — separate per-turn extraction call
- After stream end, fire a small `client.complete()` over the last K posts with a dedicated template. Model-agnostic and doesn't perturb the roleplay reply, but doubles per-turn cost/latency and needs failure tolerance (skip on `OpenRouterError`).

### Option C — retroactive backfill at absorb
- Extend `absorb.build_prompt`/`parse_output` to emit per-post state for the whole scene. Cheap (one existing call) and gives provenance, but state is never live during play — useful mainly as #121's input.

## Implementation notes
- Scene renames (`scenes.set_datetime` first-date rename, `store/scene_refs.py:repoint`) must repoint the new store — add it to the `repoint` fan-out like `changes.repoint_scenes`.
- Tolerance: garbled `turnstate.json` omits the section, never crashes `_assemble`.
- Tests: backend pytest with `monkeypatch.setenv("GRIMOIRE_HOME", tmp_path)`; parser tests for block-stripping edge cases (missing block, malformed JSON, block mid-reply).

## Dependencies & related
- #109 — extraction modes; Option A is its inline-tracker mode.
- #121 — promotion of reinforced transient state (consumer of this ledger).
- #114 — dated fact ledger; #195 — HUD would display current transient state.

### LINKED ISSUE #121: Promote reinforced transient state to canonical facts (same value across N consecutive posts)
Part of the **Extraction, Continuity & State** milestone. From the pre-rebuild target feature list:

> Promote reinforced transient state to canonical facts (same value across N consecutive posts)

## Current state
Blocked on #120 — there is no transient per-turn state layer to promote from yet. This issue defines what "promotion" means once it exists.

Canonical per-NPC state today is `backend/src/grimoire/store/playstate.py` (`characters/<cid>/state.md`, `current_state`/`knows`/`suspects`), updated exclusively through the review pipeline: `store/absorb.py:materialize` builds `character_state` StagedEdits (`{id, kind, target, label, field, before, after, authored}`), the user approves them in the absorb panel in `frontend/src/routes/CampaignView.tsx` (`saveAbsorb` → `PUT /api/campaigns/{cid}/scenes/{sid}/chronicle`), and `absorb.apply_edits` writes via `playstate.write_state`, recording browsable deltas into `changes.json`. The other "canonical" stores are the chronicle (`store/chronicle.py`, per-scene facts + `timeline.md`) — an event log, not standing state.

Note `docs/superpowers/PHASE-STATUS.md`: the established invariant is *extraction proposes, the user approves* — nothing writes canonical state silently.

### Option A — promotion as synthetic StagedEdits at absorb (recommended)
- During `POST .../absorb` (route `routes.post_absorb`), scan the scene's per-turn state ledger (#120): any `(actor, field)` holding the same normalized value across ≥N consecutive posts yields a `character_state` StagedEdit that folds the value into `current_state` (compose with `playstate.compose_body`, respecting the keep-on-omit rule in `materialize`).
- N configurable via a `store/config.py` key (e.g. `promote_streak`, string scalar, default "3").
- Rides the existing review checklist and `apply_edits` unchanged; deltas land in `changes.json` for the ChangesPanel.
- Recommended: zero new UI, preserves the review invariant, and dedupes against the LLM's own `character_state_edits` by StagedEdit id (`character_state:<cid>` — merge rather than emit twice).

### Option B — live promotion during play
- Check the streak after each turn's tracker write; on threshold, write `playstate.write_state` immediately. Canonical state updates mid-scene (better context freshness) but violates the review invariant and double-writes when absorb later re-proposes the same state.

### Option C — promote to the timeline instead
- On streak, stage a dated `timeline_events` entry (applied via `chronicle.append_timeline` in `put_chronicle`). Fits *event-shaped* reinforced facts ("adopted a limp") but standing state does not belong in an append-only log; keep as a secondary target only.

## Implementation notes
- Value equality across posts is fragile for free prose — either normalize (casefold/strip/punctuation) or constrain #120's tracked fields to short values so streaks are detectable.
- Promotion must never bypass review (Option A) and must not touch library/world files — playstate is campaign-scoped by construction.
- Tests: extend the absorb tests (`backend/tests/`, `monkeypatch.setenv("GRIMOIRE_HOME", tmp_path)`): streak met → edit staged with correct before/after; streak broken → nothing; collision with an LLM-proposed edit for the same actor → single merged edit.

## Dependencies & related
- #120 — transient per-turn state ledger (hard dependency).
- #109 — extraction modes (determines how the per-turn ledger is fed).
- #110 — confidence routing (a streak count is a natural confidence signal).
- #114 — dated fact ledger (alternative canonical target once it exists).

### PR DESCRIPTION
**Pull request by charlesmsiegel**
Closes #120. Closes #121.

Implements #120's recommended **Option A** (inline tracker block — zero extra LLM calls) and #121's **Option A** (promotion as synthetic StagedEdits at absorb). Design doc: `docs/superpowers/specs/2026-08-06-transient-turn-state-design.md`.

## What it does

The model ends each reply with a fenced `state` block:

```
{"Winifred Ash": {"mood": "guarded", "intent": "get the ledger back", "posture": "between the door and the desk"}}
```

`_persist_reply` splits that block off **before** `split_reply`, so it never enters a transcript, and files it in `/turnstate.json` keyed by `(scene, post index)` — that key is the provenance. The prompt gets the last `turnstate_depth` posts' worth back as a `Transient state` section, newest value per field; older values simply fall out of the window, which is the decay.

At absorb, an `(actor, field)` whose **final** run holds the same normalized value across `promote_streak` consecutive recorded entries yields a `character_state` StagedEdit folding `Mood: guarded` into `current_state`. It rides the existing review checklist, `apply_edits` and `changes.json` deltas unchanged — extraction proposes, the user approves. Collision with the model's own edit for that character **merges into one row** rather than opening a second: `apply_edits` writes a composed snapshot, not a patch, so two rows would mean the second erases the first.

## Off by default

`turnstate_depth = "0"`, the shape `archive_depth` already uses for "0 disables it". The tracker changes what the model is told to write on every single turn — not a change to make behind an existing install's back. Both knobs are on the Configuration page under **Transient state**.

Stripping and recording are **unconditional**; only the instruction, the injection and promotion are gated. Turning the feature off must not start leaking blocks into transcripts while the model is still complying from the scene it can see.

## Judgment calls the issues leave open

- **Promotion measures the final run, not any run.** `current_state` is standing state; a mood held four posts mid-scene and abandoned before the end is precisely what the character is *not* now.
- **An entry that does not mention an actor is skipped, not streak-breaking.** An NPC who did not act this turn has not changed her posture, and literal index adjacency makes promotion unreachable in any scene with more than one NPC.
- **Names, not store ids, key the block.** The scene prompt never shows ids. Labels resolve through `scenes.match_name` — the same rule the transcript grammar uses, so `**Winifred:**` for `Winifred Ash` works and ambiguous labels resolve to nobody.

## Keeping the ledger honest

The ledger is keyed by post index, so anything that moves or rewrites a post has to retire it. Every one of these was found by review, not by design:

- Every persisted reply supersedes entries from its landing index on (a reroll's replacement lands at the same index, so the tail filter cannot see the discarded variant's entry).
- `remove_trailing_assistant_run` supersedes from the index it **cut** at — in the mutator, not its callers, so `alternates.promote` is covered too — and parks the dropped entries in its restoration token so a failed reroll puts them back with the reply.
- `trim_continuation` supersedes from its `from_index` (preserved dice-roll lines compact *down* onto it).
- `PUT .../messages/{index}` supersedes from that index (an edit leaves the entry at a perfectly valid index).
- `delete_scene` purges the scene (ids are recycled); `scene_refs.repoint` follows renames.
- Promotion measures the ledger **snapshotted with the scene** under one lock before the extraction call, so nothing landing mid-absorb reaches a review that never saw it.

## Other seams

- **In-stream suppression.** A `StreamRedactor` composed *outside* `fence.FenceWatcher`. It withholds only at a line boundary (what the block grammar requires) and resolves at end of stream with `split_block` itself, so the stream and the transcript cannot disagree. `watcher.narration` is untouched.
- **Only a trailing block is stripped.** A fence with narration after it stays put — deleting from the middle of a reply is the one failure here that loses story.
- **Every "did this turn produce anything?" test runs on the stripped narration**, so a tracker-only reply counts as empty and the reroll restore fires.
- Tracker values are macro-expanded once at persist time, like narration.
- `store.turnstate` is classified in `locks.DOMAIN_MODULES`; a scene keeps its newest 200 entries.
- The tracker instruction never renders on an opener (`Section.except_opener`), and is omitted on a turn that requests a dice check (the roll block ends the turn and takes precedence).
- Context assembly pairs the transcript with the ledger under `locks.best_effort_campaign_lock` — a *read* path that must never 409, because `post_chat` has already appended the player's post by then.

## Review

Nine rounds of Codex review on this PR, which found **twenty-five** defects — including one that permanently destroyed a reply a reroll had parked, one where CRLF line endings silently disabled the whole feature, one where `commit_narration` marked a proposal `narrated` with nothing landed, and one regression introduced by an earlier round's fix. Twenty-four are fixed, each with a regression test verified to fail without its fix (revert the fix, watch the test go red).

**One finding is declined, and its thread is left open deliberately** so it can be overruled: the suggestion to clamp `promote_streak` to `MAX_ENTRIES`. `turnstate_depth` counts *posts* and `MAX_ENTRIES` counts *entries*, and entries are sparse — a scene can hold far more than 200 posts' worth of window in under 200 entries. The clamp is a no-op in the dense case and discards retained, in-window data in the sparse case. A mutation check confirmed the proposed clamp was unreachable by the test written for it. Two tests now pin both behaviours, and `current()` documents the unit mismatch.

An earlier version of this description claimed three rounds of a local `brutal-review` substitute had covered this; that was written before Codex ran and substantially understated what review found. The `CLAUDE.md` **spec-conformance** gate (`/codex:adversarial-review` against the diff *and* the originating spec, asking whether this implements #120/#121 rather than whether the diff is clean) has still not been run in that form.

## Testing

Rebased onto `main` (2972cc6), which brought in the #157 prompt-log feature and the #194 setup wizard. Full gate re-run locally after the rebase:

- `pytest backend` — **4069 passed**
- `ruff` — clean
- `verify_templates.py` — 62 checks, builders and `templates/` agree byte-for-byte
- frontend `tsc --noEmit` + `vitest run` — **918 passed** across 60 files

One pre-existing failure is unrelated and also fails on `main`: `test_atomic.py::test_a_read_only_record_is_not_silently_replaced` needs a non-root user to make a file read-only.

**CI is green on the rebased head** — all seven jobs (lint, templates, frontend, backend py3.11, backend py3.14, pydantic 1.10 / Android dependency set, android debug apk) passed on run 31139565232. This is the first `pull_request` run this branch has had; earlier revisions were verified only by `make check` locally.

Not covered: whether a live model actually follows the tracker instruction — only `evals/run.py --live` answers that.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01PQtcfFndTaBTC5Ba9kogH7

### CODE REVIEW

File: backend/src/grimoire/store/turnstate.py [hunk 1]
```diff
@@ -0,0 +1,577 @@
...
+    mysteries."""
+    numbered = []
+    for key in scene:
+        try:
+            numbered.append((int(key), key))
+        except (TypeError, ValueError):
+            continue
+    if len(numbered) <= MAX_ENTRIES:
+        return scene
+    drop = {key for _, key in sorted(numbered)[:len(numbered) - MAX_ENTRIES]}
```
[the remark this answers is not in the document]
Comment (charlesmsiegel) [reply backend/src/grimoire/store/turnstate.py:434]: Not taking this one — I tried it, and my own mutation check caught the fix passing with the clamp removed. Working out why showed the suggestion is wrong here.

The two bounds are in **different units**. `depth` counts *posts* back from the tail; `MAX_ENTRIES` counts *recorded entries*, and entries are sparse — a scene where one NPC was tracked at post 0 and another at post 1000 holds exactly two of them.

- In the dense case you describe (201 single-post replies), index 0 has **already been evicted by the cap**. `current()` cannot show it whatever the floor is, so clamping changes nothing observable.
- In the sparse case, clamping actively *loses* data: with entries at posts 0 and 1000 and `turnstate_depth = 1001`, the post-0 actor is inside the requested window **and still retained**, and clamping the floor to `tail - 200` drops her.

So the clamp is a no-op exactly where it was aimed and a regression where it bites. The other option you offer — retaining enough entries to satisfy the configured depth — would make retention config-driven and remove the file-growth bound that `MAX_ENTRIES` exists to provide, since the setting is unbounded.

What I did instead: documented the relationship at `current()` — the window is a ceiling on what is asked for, the cap is the memory that answers, and where they disagree the answer is whatever survived — and pinned **both** behaviours with tests (`test_a_wide_window_still_shows_every_entry_the_cap_kept`, `test_a_narrow_window_still_decays`), so re-introducing the clamp now fails the suite.

`promote_streak` *is* still clamped, and that one is right: it counts consecutive entries, so it is in the cap's own units.

---
_Generated by [Claude Code](https://claude.ai/code)_
